### PR TITLE
WebTransport

### PR DESCRIPTION
**Pull request by pthatcherg**
Here's an idea for breaking up QuicTransport into UnidirectionalStreamsTransport, BidirectionalStreamsTransport, and DatagramTransport to allow for other protocol (H2? H3?) to be used underneath the same more generic APIs.

It's a big change, so lots of discussion to be had.  This PR can serve as a concrete thing to look at and discuss.

### CODE REVIEW

File: cs.html [hunk 1]
```diff
@@ -75,288 +81,110 @@ <h2>Terminology</h2>
...
-    readonly        attribute QuicTransportState    state;
-    void                  stop (QuicTransportStopInfo stopInfo);
-    Promise&lt;QuicBidirectionalStream&gt;         createBidirectionalStream ();
-    Promise&lt;QuicSendStream&gt;                  createSendStream (optional QuicStreamParameters parameters);
-                    attribute EventHandler             onstatechange;
-                    attribute EventHandler             onerror;
-                    attribute EventHandler             onreceivestream;
-                    attribute EventHandler             onbidirectionalstream;
+    <pre class="idl">
+interface UnidirectionalStreamsTransport {
```
Comment (pthatcherg) [inline cs.html]: interface mixin

File: cs.html [hunk 2]
```diff
@@ -446,35 +274,158 @@ <h2>Attributes</h2>
...
+      Data within a stream is delivered in order, but data between streams may be delivered out of order.
+      Data is generally sent reliably, but retransmissions may be disabled 
+      or the stream may aborted to produce a form of unreliability.
+      All stream data is encrypted and congestion-controlled.
+    </p>
+    <pre class="idl">
+interface BidirectionalStreamsTransport {
+    Promise&lt;BidirectionalStream&gt; createBidirectionalStream ();
+    attribute EventHandler             onbidirectionalstream;
+};</pre>
```
Comment (pthatcherg) [inline cs.html]: interface mixin

File: cs.html [hunk 3]
```diff
@@ -528,258 +479,278 @@ <h2>Attributes</h2>
...
+  <section id="datagram-transport*">
+    <h2><dfn>DatagramTransport</dfn> Mixin</h2>    
+    <p>
+      A <code>DatagramTransport</code> can send and receive datagrams.
+      Datagrams are sent out of order, unreliably, and have a limited maximum size.
+      Datagrams are encrypted and congestion controlled.
+      </p>
+      <pre class="idl">
+interface DatagramTransport {
+    readonly attribute unsigned short         maxDatagramSize;
```
Comment (pthatcherg) [inline cs.html]: interface mixin

File: cs.html [hunk 4]
```diff
@@ -791,412 +762,444 @@ <h2>Constructors</h2>
...
-      a <code><a>QuicTransportBase</a></code> instance.</p>
+
+  <section id="data-transport*">
+    <h2><dfn>DataTransport</dfn> Mixin</h2>   
+    <p>
+      The <code>DataTransport</code> includes the methods common to all data transports,
+      such as state, state changes, and the ability to stop the transport.
+    </p>
+    <pre class="idl">
+interface DataTransport {
```
Comment (pthatcherg) [inline cs.html]: interface mixin

**Comment by pthatcherg** [comment]
Note from editor's call: we like the general approach and the idea of taking the API to the WICG with this model.  Clearly needs PR review :)

**Comment by pthatcherg** [comment]
Note from editor's call: we like the general approach and the idea of taking the API to the WICG with this model.  Clearly needs PR review :), especially around:

- slots
- whether we should have DataTransport mixin or not 
- If the "For QUIC, ..." type sentences make sense where they are.

And if someone wants to make a DataTransport based on H2/H3, that would be cool too.

**Comment by shampson** [comment]
Easy viewing:
https://rawgit.com/w3c/webrtc-quic/web-transport/index.html

**Comment by lgrahl** [comment]
I'm not sure at what kind of *level* this should be discussed.

Abstracting web transports and generalising their APIs is an idea I very much support.

The use of this concrete API is something where I would repeat myself and point to #2.

File: cs.html [hunk 5]
```diff
@@ -961,689 +1180,684 @@ <h2>Constructors</h2>
...
+          </dd>
+        </dl>
+      </section>
+      <section>
+        <h3>Methods</h3>
+        <dl data-link-for="OutgoingStream" data-dfn-for="OutgoingStream" class=
+        "methods">
+          <dt><dfn><code>write</code></dfn></dt>
+          <dd>
+            <p>Writes data to the stream. When the remote <code><a>DataTransport</a></code>
```
Comment (vasilvv) [inline cs.html]: Maybe it would be more clear to say it "adds data to the write buffer"?  I feel like that's the best summary of what happens, as well as to why this method returns void.

File: cs.html [hunk 6]
```diff
@@ -961,689 +1180,684 @@ <h2>Constructors</h2>
...
+            <p>The <dfn id="dom-incomingstream-readableamount"><code>readableAmount</code></dfn>
+            attribute represents the number of bytes buffered for access by
+            <code>readInto</code> but that, as of the last time the event loop
+            started executing a task, had not yet been read. This does not include
+            framing overhead incurred by the protocol, or buffers associated with
+            the network hardware. On getting, it <em class="rfc2119" title="MUST">MUST</em>
+            return the value of the <code><a>IncomingStream</a></code>'s
+            <a>[[\ReadableAmount]]</a> internal slot.</p>
+          </dd>
+          <dt><code>readingAborted</code> of type <span class="idlAttriType"><a>StreamAbortInfo</a>
```
Comment (vasilvv) [inline cs.html]: idlAttrType?  Also I think the span tag should be closed here.

File: cs.html [hunk 7]
```diff
@@ -961,689 +1180,684 @@ <h2>Constructors</h2>
...
+                <tr>
+                  <th>Parameter</th>
+                  <th>Type</th>
+                  <th>Nullable</th>
+                  <th>Optional</th>
+                  <th>Description</th>
+                </tr>
+                <tr>
+                  <td class="prmName">data</td>
+                  <td class="prmType"><code>Uint8Array</code></td>
```
Comment (vasilvv) [inline cs.html]: I'm not entirely familiar with JavaScript binary array type system, so just wanted to check: is it possible, with current method signature, to write at an arbitrary offset in some array? I.e. can I pass a view instead of the entire array?

File: cs.html [hunk 8]
```diff
@@ -1674,32 +1889,38 @@ <h2>Event summary</h2>
...
           sending data on it. Prior to <code><a>bidirectionalstream</a></code>
-          firing, the <code><a>QuicBidirectionalStream</a></code> is added to the
-          <code><a>QuicTransportBase</a></code>'s <a>[[\QuicTransportReadableStreams]]</a>
-          and <a>[[\QuicTransportWritableStreams]]</a> internal slots.</td>
+          firing, the <code><a>BidirectionalStream</a></code> is added to the
+          <code><a>DataTransport</a></code>'s <a>[[\IncomingStreams]]</a>
+          and <a>[[\OutgoingStreams]]</a> internal slots.</td>
+        </tr>
+        <tr>
+          <td><dfn><code>datagramsreceived</code></dfn></td>
```
Comment (vasilvv) [inline cs.html]: I think this is a duplicate of the event below (you might want to delete both of them though, since they aren't used anywhere currently).

File: cs.html [hunk 9]
```diff
@@ -1710,60 +1931,61 @@ <h2>Event summary</h2>
...
+const transport = getTransport();
+setInterval(() => {
+  transport.sendDatagram(createDatagram());
+}, 100);</pre>
+    </section>
+
+    <section class="informative" id="datagramexample3*">
+      <h3>Receiving datagrams</h3>
+      <p>Receiving datagrams can be achieved by simply listening to the
       <code>datagramsreceived</code>, remembering to check for null values
```
Comment (vasilvv) [inline cs.html]: datagramsreceived is not defined in the current text.

File: cs.html [hunk 10]
```diff
@@ -75,437 +81,110 @@ <h2>Terminology</h2>
...
-      for <code>QuicTransport</code>.  Most of the functionality of a
-      QuicTransport is in the base class to allow for other
-      subclasses (such as a p2p variant) to share the same interface.
+
+  <section id="unidirectional-streams-transport*">
+    <h2><dfn>UnidirectionalStreamsTransport</dfn> Mixin</h2>    
+    <p>
+      A <code>UnidirectionalStreamsTransport</code> can send an receive unidirectional streams.
+      Data within a stream is delivered in order, but data between streams may be delivered out of order.
+      Data is generally sent reliably, but retransmissions may be disabled
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: It's something that's been in the spec for a while.  In Chromium we can support it be canceling a stream when we detect loss using this: https://cs.chromium.org/chromium/src/net/third_party/quic/quartc/quartc_stream.h?l=69.

The only extra QUIC API you need is to know when there is loss, which we already have in Chromium's QUIC impl.

File: cs.html [hunk 11]
```diff
@@ -519,16 +198,16 @@ <h4 id="add-send-stream-to-transport">Add QuicSendStream to the QuicTransportBas
...
-      <h3><dfn>QuicStreamParameters</dfn> Dictionary</h3>
+      <h3><dfn>SendStreamParameters</dfn> Dictionary</h3>
       <p>The <code>QuicStreamParameters</code> dictionary includes information
-      relating to QUIC stream configuration.</p>
+      relating to stream configuration.</p>
       <div>
-        <pre class="idl">dictionary QuicStreamParameters {
-             bool disableRetransmissions = false;
+        <pre class="idl">dictionary SendStreamParameters {
+              bool disableRetransmissions = false;
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: Good question.  Some options:

A.  Silently still be reliable, which is the same as if QUIC is running on ICE running on TCP.

B.  Throw an exception.

C.  Add something to the WebTransport to indicate what it supports (and also do A or B).

D.  Assume the developers knows what kind of transport it is and what it supports.  


I'm inclined toward C+A at the moment.

File: cs.html [hunk 12]
```diff
@@ -75,437 +81,110 @@ <h2>Terminology</h2>
...
+              <p>Let <var>transport</var> be the <code><a>UnidirectionalStreamsTransport</a></code>
+              on which <code>createSendStream</code> is invoked.</p>
+            </li>
+            <li>
+              <p>If <code><var>transport</var>'s state</code> is <code>"closed"</code> or
+              <code>"failed"</code>, immediately return a new <a>rejected</a> promise with a
+              newly created <code>InvalidStateError</code> and abort these steps.</p>
+            </li>
+            <li>
+              <p>If <code><var>transport</var>'s state</code> is <code>"connected"</code>,
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: I'll put a TODO to change this for if/when we support 0-RTT.

File: cs.html [hunk 13]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+        <dd>
+          <p>The maximum size data that may be passed to sendDatagram.</p>
+        </dd>
+      </dl>
+    </section>
+    <section>
+      <h2>Methods</h2>
+      <dl data-link-for="DatagramTransport" data-dfn-for="DatagramTransport" class=
+      "methods">
+        <dt><dfn><code>readyToSendDatagram</code></dfn></dt>
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: readyToSendDatagrams won't block until there are many datagrams in a queue, so I don't think a datagram sender will starve, unless the streams are higher priority than the datagrams.  We haven't specified datagrams vs. streams priority, but it doesn't make sense to prioritize datagrams over streams.

File: cs.html [hunk 14]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+            <li>
+              <p>Return <var>p</var> and continue the following steps in
+              the background.</p>
+            </li>
+            <ol>
+              <li>
+                <p>Send <var>data</var> in a datagram.</p>
+              </li>
+              <li>
+                <p>When an ack is received for the sent datagram,
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: That's correct, the ACK can be deceiving.   On the other hand, without it there isn't a good way for datagram senders to know they need to resend, short of adding another layer of ACKs by the application.

File: cs.html [hunk 15]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+		         If a promised returned from a previous call is still unresolved, 
+            the user agent MUST return a new promise rejected with an InvalidStateError. </p>   
+          <div>
+            <em>Return type:</em> <code>Promise&lt;sequence&lt;Uint8Array&gt;&gt;</code>
+          </div>
+        </dd>
+      </dl>
+    </section>
+    <section>
+      <h3><dfn>DatagramsReceivedEvent</dfn></h3>
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: Yes, I forgot to remove it.  Removing again now.

File: cs.html [hunk 16]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+                <p>The payloads of the datagrams received.</p>
+            <dd>
+          </dl>
+        </section>
+      </div>
+    </section>
+  </section>
+
+  <section id="data-transport*">
+    <h2><dfn>DataTransport</dfn> Mixin</h2>
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: I like that idea.  Done

File: cs.html [hunk 17]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+                <code>stop()</code> or receipt of a closing message from the remote side.
+                When the <code><a>DataTransport</a></code>'s
+                internal <a>[[\TransportState]]</a> slot transitions to
                 <code>closed</code> the user agent <em class="rfc2119" title="MUST">MUST</em>
                 run the following steps:</p>
                 <ol>
-                  <li>Let <var>transport</var> be the <code><a>QuicTransportBase</a></code>.
-                  <li>For each <code><a>QuicReadableStream</a></code> in <var>transport</var>'s
-                  <a>[[\QuicTransportReadableStreams]]</a> internal slot run the
+                  <li>Let <var>transport</var> be the <code><a>DataTransportBase</a></code>.
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: WebTransport :)

File: cs.html [hunk 18]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+  <section id="data-transport*">
+    <h2><dfn>DataTransport</dfn> Mixin</h2>   
+    <p>
+      The <code>DataTransport</code> includes the methods common to all data transports,
+      such as state, state changes, and the ability to stop the transport.
+    </p>
+    <pre class="idl">
+interface mixin DataTransport {
+  readonly attribute TransportState state;
+  void                                  stop (DataTransportStopInfo stopInfo);
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: Done

File: cs.html (hunk 9 above)
Comment (pthatcherg) [reply cs.html]: Removed

File: cs.html (hunk 8 above)
Comment (pthatcherg) [reply cs.html]: Yep, both removed.

File: cs.html (hunk 7 above)
Comment (pthatcherg) [reply cs.html]: Yes, that's what a Uint8Array is: it's basically a view of an ArrayBuffer.

File: cs.html (hunk 6 above)
Comment (pthatcherg) [reply cs.html]: Good find.  Fixed.

File: cs.html (hunk 5 above)
Comment (pthatcherg) [reply cs.html]: Good idea.

File: cs.html (hunk 11 above)
[the remark this answers is not in the document]
Comment (vasilvv) [reply cs.html]: I feel like C+A should be the right solution (to make switching transport transparently easy), so I'd do A in this PR, and file C as an issue.

[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: OK, A added in this CR with TODO for C.

File: cs.html (hunk 14 above)
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: I think you're right.  I've filed on issue for that: https://github.com/w3c/webrtc-quic/issues/117

File: cs.html [hunk 19]
```diff
@@ -677,147 +479,519 @@ <h2>Attributes</h2>
...
+            </li>
+            <li>
+              <p>If <code><var>data</var></code> is too large to fit into a
+              datagram, return a promise <a>rejected</a> with a newly created
+              <code>InvalidArgumentError</code> and abort these steps.</p>
+            </li>
+            <li>
+              <p>If <var>transport</var> is unable to send the datagram due
+              to congestion control, return a promise <a>rejected</a> with
+              a newly created <code>InvalidStateError</code> and abort
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: Made an issue for that too: https://github.com/w3c/webrtc-quic/issues/118

File: cs.html [hunk 20]
```diff
@@ -961,689 +1180,684 @@ <h2>Constructors</h2>
...
+    Promise&lt;void&gt; waitForWriteBufferedAmountBelow(unsigned long threshold);
+};</pre>
+      <section>
+        <h3>Overview</h3>
+        <p>The <code><a>OutgoingStream</a></code> will initialize with
+        the following:</p>
+        <ol>
+          <li>Let <var>stream</var> be the <code><a>OutgoingStream</a></code>.</li>
+          <li>Let <var>stream</var> have a <dfn>[[\Writable]]</dfn> internal
+          slot initialized to <code>true</code>.</li>
```
[the remark this answers is not in the document]
Comment (pthatcherg) [reply cs.html]: An outgoing stream is writable when it's given to the app (starts off writable)

**Comment by shampson** [comment]
https://rawgit.com/w3c/webrtc-quic/web-transport/cs.html
If you load the page there's some errors/warnings:
No <dfn> for attribute streamId in WebTransportStream. More info.
No <dfn> for attribute transport in WebTransportStream. More info.
No <dfn> for interface BidirectionalStream. More info.
No <dfn> for interface ReceiveStream. More info.
Found linkless element for 'webtransportstate' with text 'datratransport/a>' but no matching <dfn>.
Found linkless element for 'webtransportstate' with text ''s internal' but no matching <dfn>.
Found linkless element with text '[[quictransportstate]]' but no matching <dfn>.
Found linkless element with text '[[quictransportstate]]' but no matching <dfn>.
Found linkless element with text '[[quictransportstate]]' but no matching <dfn>.
Found linkless element with text '[[quictransportstate]]' but no matching <dfn>.
Found linkless element with text 'quicstream' but no matching <dfn>.

File: cs.html [hunk 21]
```diff
@@ -961,689 +1090,685 @@ <h2>Constructors</h2>
...
-              method is called, the user agent <em class="rfc2119" title="MUST">MUST</em>
-              run the following steps:</p>
+          </dl>
+        </section>
+      </div>
+    </section>
+  </section>
+
+  <section id="outgoing-stream*">
+    <h2>Interface Mixin <dfn>OutgoingStream</dfn></h2>
```
Comment (shampson) [inline cs.html]: Why the change from WritableStream to OutgoingStream?

Comment (shampson) [reply cs.html]: and with ReadableStream as well.

File: cs.html [hunk 22]
```diff
@@ -961,689 +1090,685 @@ <h2>Constructors</h2>
...
+          </section>
+        </div>
+      </section>
+    </div>
+  </section>
+
+  <section id="incoming-stream*">
+    <h2>Interface Mixin <dfn>IncomingStream</dfn></h2>
+    <p>
+        An OutgoingStream is a stream that can be read from,
```
Comment (shampson) [inline cs.html]: "An OutgoingStream" --> An Incomingstream

File: cs.html (hunk 21 above)
Comment (pthatcherg) [reply cs.html]: To avoid name conflicts with WHATWG streams (which we may use in the future).

File: cs.html (hunk 22 above)
Comment (pthatcherg) [reply cs.html]: Fixed

**Comment by pthatcherg** [comment]
I have fixed all the respec errors except for the ones about linking to Bidirectiont,alStream and ReceiveStream.    I haven't figured out how to fix those yet.

File: cs.html (hunk 21 above)
Comment (shampson) [reply cs.html]: Got it.

File: cs.html (hunk 2 above)
Comment (shampson) [reply cs.html]: This is still missing mixin.

File: cs.html [hunk 23]
```diff
@@ -75,468 +81,147 @@ <h2>Terminology</h2>
...
+            <li>
+              <p>Let <var>p</var> be a new promise.</p>
+            </li>
+            <li>
+              <p>Return <var>p</var> and continue the following steps in
+              the background.</p>
+            </li>
+            <ol>
+              <li>
+                <p>When <code><var>transport</var>'s state</code> transitions to
```
Comment (shampson) [inline cs.html]: I think you'll need to merge in (or copy over) the PR we merged last week that updates this to include waiting on congestion control as well. Here and also in createBidirectionalStream.

PR that merged: https://github.com/w3c/webrtc-quic/pull/120

File: cs.html [hunk 24]
```diff
@@ -961,689 +1090,685 @@ <h2>Constructors</h2>
...
+            <dl data-link-for="StreamReadResult" data-dfn-for="StreamReadResult" class=
+            "dictionary-members">
+              <dt><dfn><code>amount</code></dfn> of type <span class=
+              "idlMemberType"><a>unsigned long</a></span>.</dt>
+              <dd>
+                <p>The amount of data read in bytes.</p>
+              </dd>
+              <dt><dfn><code>finished</code></dfn> of type <span class=
+              "idlMemberType">boolean</span>.</dt>
+
```
Comment (shampson) [inline cs.html]: Could change to QuicReceiveStream and QuicBidirectionalStream to avoid the name conflict?

File: cs.html [hunk 25]
```diff
@@ -961,689 +1090,685 @@ <h2>Constructors</h2>
...
+      [ Exposed=Window ]
+interface mixin OutgoingStream {
+    readonly attribute boolean writable;
+    readonly attribute unsigned long writeBufferedAmount;
+    readonly attribute Promise&lt;StreamAbortInfo&gt; writingAborted;
+    void write (StreamWriteParameters data);
+    void abortWriting (StreamAbortInfo abortInfo);
+    Promise&lt;void&gt; waitForWriteBufferedAmountBelow(unsigned long threshold);
+};</pre>
+      <section>
```
Comment (shampson) [inline cs.html]: For some reason "Interface Mixin OutgoingStream" & the "Interface Mixin IncomingStream" aren't split up into subsections appropriately like in Transport Mixins for example:

https://rawgit.com/w3c/webrtc-quic/web-transport/cs.html#outgoing-stream*

I'm assuming it's an html issue, missing a closing element or something related.

File: cs.html (hunk 24 above)
Comment (pthatcherg) [reply cs.html]: I was trying to have the stream names apply to non-QUIC transports as well.

File: cs.html (hunk 23 above)
Comment (pthatcherg) [reply cs.html]: Done

**Comment by pthatcherg** [comment]
I renamed cs.html to webtransport.html in this PR.

**Comment by shampson** [comment]
There's still a couple warnings/errors and the html formatting issue I mentioned above, but we can follow up on this stuff:
https://w3c.github.io/webrtc-quic/webtransport.html
